### PR TITLE
Relax `tar` upper bound

### LINKED ISSUE #9556: Support tar-0.6
`cabal-install` demands `tar < 0.6`, prohibiting a recently released https://hackage.haskell.org/package/tar-0.6.0.0. 
Cf. https://github.com/commercialhaskell/stackage/issues/7202

### PR DESCRIPTION
**Pull request by ffaf1**
**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Close #9556 

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

### CODE REVIEW

**Comment by Bodigrim** [comment]
@ffaf1 you also need to update bounds in `Cabal-tests.cabal`, otherwise CI would not pick up `tar-0.6`.

**Review by ulysses4ever — APPROVED** [review]

File: cabal-install/src/Distribution/Client/Tar.hs [hunk 1]
```diff
@@ -69,8 +69,12 @@ extractTarGzFile
...
   -> IO ()
 extractTarGzFile dir expected tar =
-  Tar.unpack dir
-    . Tar.checkTarbomb expected
+  Tar.unpackAndCheck
+    ( \x ->
+        SomeException <$> Tar.checkEntryTarbomb expected x
+          <|> SomeException <$> Tar.checkEntrySecurity x
+    )
+    dir
```
Comment (Bodigrim) [inline cabal-install/src/Distribution/Client/Tar.hs:77]: `unpackAndCheck` is a new API, unavailable from `tar-0.5` series. You can either bump lower bound to `tar >= 0.6`, or switch between `unpack` and `unpackAndCheck` with CPP, checking `#if MIN_VERSION_tar(0,6,0)`.

Comment (gbaz) [reply cabal-install/src/Distribution/Client/Tar.hs:77]: I think the approach in keeping with Cabal practices would be to create Distribution.Compat.Tar and use CPP there, creating a helper function which has the same type across both versions.

Comment (gbaz) [reply cabal-install/src/Distribution/Client/Tar.hs:77]: That said, could we not just call `checkSecurity` and `checkTarbomb` in succession which would let us keep making use of the old 0.5 apis that 0.6 still supports, and thus avoid the need for CPP altogether?

Comment (Bodigrim) [reply cabal-install/src/Distribution/Client/Tar.hs:77]: @gbaz No, types are different.

**Review by Bodigrim — APPROVED** [review]

**Comment by ffaf1** [comment]
I need help with CI. 

```
<no location info>: error: [-Wunused-packages, -Werror=unused-packages]
    The following packages were specified via -package or -package-id flags,
    but were not needed for compilation:
      - tar-0.6.0.0
```

**Comment by geekosaur** [comment]
Looks to me like it's only used in a test, but you added it to cabal-install?

**Comment by geekosaur** [comment]
Come to think of it, that can't be right because it is used to read the package index. Which means there's at least one missing change somewhere.

**Comment by Bodigrim** [comment]
This is likely a bug on GHC side caused by an internal sublibrary, please disable the warning.

**Comment by Bodigrim** [comment]
@gbaz is it good to go?

**Review by gbaz — APPROVED** [review]

**Comment by ffaf1** [comment]
@mergify backport

**Comment by ffaf1** [comment]
@mergify backport 3.10

**Pull request merged**